File: Programming and Projection Methods/Lab10/Laborator10Java/CompanyPersistance/src/main/java/company/persistance/ClientRepository.java
package company.persistance;


import company.model.Clientj;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;


import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Properties;

public class ClientRepository implements IClientRepository {
    private DBConnection dbConnection;
    private static final Logger logger = LogManager.getLogger();
    private Validator<Clientj> valid;

    public ClientRepository(Properties properties, ClientValidator v) {
        logger.info("Initializing ClientRepository with properties: {}", properties);
        dbConnection = new DBConnection(properties);
        this.valid = v;
    }

    @Override
    public Clientj findOne(Integer integer) {
        return null;
    }

    public Clientj findOne(String name) {
        logger.traceEntry("Finding client with name {}", name);
        Connection conn = dbConnection.getConnection();
        try (PreparedStatement stat = conn.prepareStatement("select * from Client where name=?")) {
            stat.setString(1, name);
            try (ResultSet resultSet = stat.executeQuery()) {
                resultSet.next();
                int id = resultSet.getInt("client_id");
                Clientj c = new Clientj(id, name);
                return c;
            }
        } catch (SQLException e) {
            logger.error(e);
            throw new RepositoryException("Ride couldn't be found in the database!");
        }
    }

    @Override
    public void save(Clientj client) throws ValidationException {
        logger.traceEntry("Saving client {} ", client);
        valid.validate(client);
        Connection conn = dbConnection.getConnection();
        try (PreparedStatement stat = conn.prepareStatement("INSERT INTO Client (name) VALUES (?)")) {
            stat.setString(1, client.getName());
            stat.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            throw new RepositoryException("Database error " + e);
        }
        logger.traceExit();
    }

    public Clientj findLastAdded() {
        logger.traceEntry("Finding last client added");
        Connection conn = dbConnection.getConnection();
        try (PreparedStatement stat = conn.prepareStatement("select max(client_id) as [id],name from Client")) {
            try (ResultSet resultSet = stat.executeQuery()) {
                resultSet.next();
                Integer id = resultSet.getInt("id");
                String name = resultSet.getString("name");
                Clientj c = new Clientj(id, name);
                return c;
            }
        } catch (SQLException e) {
            logger.error(e);
            throw new RepositoryException("Client couldn't be found in the database!");
        }
    }

    @Override
    public void update(Clientj client) {

    }

    @Override
    public int size() {
        return 0;
    }

    @Override
    public Iterable<Clientj> findAll() {
        return null;
    }

    @Override
    public void delete(Integer integer) {

    }
}

File: Programming and Projection Methods/Lab8C#/Laborator8/CompanyPersistance/ClientRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using CompanyModel;
using log4net;

namespace CompanyPersistance
{
    public class ClientRepository : IClientRepository
    {
        private static readonly ILog log = LogManager.GetLogger("RideRepository");
        IDictionary<String, String> props;
        public ClientRepository(IDictionary<String, String> props)
        {
            log.Info("Creating ClientRepository");
            this.props = props;
        }

        public Client FindOne(String name)
        {
            log.InfoFormat("Finding client with name {0}", name);
            IDbConnection con = DBUtils.getConnection(props);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select * from Client where name=@name";
                IDbDataParameter paramName = comm.CreateParameter();
                paramName.ParameterName = "@name";
                paramName.Value = name;
                comm.Parameters.Add(paramName);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        int id = dataR.GetInt32(0);
                        Client c = new Client(id, name);
                        return c;
                    }
                }
            }
            throw new RepositoryException("Client nu a fost gasit!");
        }


        public void Delete(int id)
        {
            throw new NotImplementedException();
        }

        public IEnumerable<Client> FindAll()
        {
            throw new NotImplementedException();
        }

        public Client FindOne(int id)
        {
            throw new NotImplementedException();
        }

        public void Save(Client e)
        {
            log.InfoFormat("Saving client {0}", e);
            IDbConnection con = DBUtils.getConnection(props);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "insert into Client (name) values (@name)";
                IDbDataParameter paramName = comm.CreateParameter();
                paramName.ParameterName = "@name";
                paramName.Value = e.Name;
                comm.Parameters.Add(paramName);
                var result = comm.ExecuteNonQuery();
                if (result == 0)
                    throw new RepositoryException("No client added !");

                log.InfoFormat("Client inserted successfully");
            }
        }

        public Client FindLastAdded()
        {
            log.InfoFormat("Finding last client added");
            IDbConnection con = DBUtils.getConnection(props);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select max(client_id),name from Client";
                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        int id = dataR.GetInt32(0);
                        String name = dataR.GetString(1);
                        Client c = new Client(id, name);
                        return c;
                    }
                    log.ErrorFormat("Client couldn't be found in the database!");
                    throw new RepositoryException("Client cound't be set!");
                }
            }
        }

        public void Update(Client e)
        {
            throw new NotImplementedException();
        }
    }
}
